### ISSUE TITLE
Random CI failures with packages that depend on omake

### ISSUE THREAD

**Issue by rgrinberg** [ocaml-omake/omake#71]
See these builds for example:

https://travis-ci.org/ocaml/opam-repository/builds/180491542

4.03 seems to fail on OSX/Linux with the following error:

```
#=== ERROR while installing ppx_test.1.4.0 ====================================#
# opam-version 1.2.2
# os           linux
# command      ocaml setup.ml -build
# path         /home/travis/.opam/4.03.0/build/ppx_test.1.4.0
# compiler     4.03.0
# exit-code    1
# env-file     /home/travis/.opam/4.03.0/build/ppx_test.1.4.0/ppx_test-14362-e80101.env
# stdout-file  /home/travis/.opam/4.03.0/build/ppx_test.1.4.0/ppx_test-14362-e80101.out
# stderr-file  /home/travis/.opam/4.03.0/build/ppx_test.1.4.0/ppx_test-14362-e80101.err
### stdout ###
# [...]
# - build ppx ppx_test.o
# + ocamlfind ocamlc -annot -bin-annot -package compiler-libs.common,ppxx -g -w A-3-4-9-40-42-44-45-48 -warn-error a -g -I . -c ppx_test.ml
# - build ppx ppx_test.o
# + ocamlfind ocamlopt -annot -bin-annot -package compiler-libs.common,ppxx -g -w A-3-4-9-40-42-44-45-48 -warn-error a -I . -c ppx_test.ml
# - build ppx ppx_test.opt
# + ocamlfind ocamlopt -package compiler-libs.common,ppxx -g -I . -o ppx_test.opt ppx_test.cmx -linkpkg
# - build ppx ppx_test.run
# + ocamlfind ocamlc -package compiler-libs.common,ppxx -g -g -I . -o ppx_test.run ppx_test.cmo -linkpkg
# *** omake: 64/101 targets are up to date
# *** omake: failed (0.63 sec, 7/7 scans, 15/23 rules, 47/132 digests)
### stderr ###
# [...]
# findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /home/travis/.opam/4.03.0/lib/ocaml, /home/travis/.opam/4.03.0/lib/ocaml/compiler-libs
# findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /home/travis/.opam/4.03.0/lib/ocaml, /home/travis/.opam/4.03.0/lib/ocaml/compiler-libs
# findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /home/travis/.opam/4.03.0/lib/ocaml, /home/travis/.opam/4.03.0/lib/ocaml/compiler-libs
# File "_none_", line 1:
# Error: Files ppx_test.cmo and /home/travis/.opam/4.03.0/lib/ppxx/ppxx.cmo
#        make inconsistent assumptions over interface Ppxx
# *** omake: targets were not rebuilt because of errors:
#    ppx/ppx_test.run
#       depends on: ppx
# E: Failure("Command 'sh -c \"cp OMakeroot.in OMakeroot; omake PREFIX=/home/travis/.opam/4.03.0\"' terminated with error code 2")
```

I'm at a loss as to what could be causing this and why could this possibly be 4.03 specific. Would like to debug this somehow but I have no idea how to reproduce.

**Comment by gerdstolpmann** [comment]
Usually such errors mean that the dependency graph is incomplete (and because omake is usually running two compiler commands in parallel this leads then to corrupted files). Lately I could identify one such reason: https://github.com/ocaml-omake/omake/issues/73.

Note that `orakuda` uses its own build rules (`OMyMakefile`) which makes it more difficult to analyze. I cannot say whether these are buggy or not.

**Comment by yallop** [comment]
The problem doesn't seem to be 4.03-specific.  [Here](https://travis-ci.org/ocaml/opam-repository/jobs/177269550)'s an example of a similar failure on 4.04.0:

```
# File "_none_", line 1:
# Error: Files ppx_test.cmo and /home/travis/.opam/4.04.0/lib/ppxx/ppxx.cmo
#        make inconsistent assumptions over interface Ppxx
# *** omake: targets were not rebuilt because of errors:
#    ppx/ppx_test.run
#       depends on: ppx
# E: Failure("Command 'sh -c \"cp OMakeroot.in OMakeroot; omake PREFIX=/home/travis/.opam/4.04.0\"' terminated with error code 2")
```

**Comment by gerdstolpmann** [comment]
Both of these projects (orakuda, ppx_meta:conv) have in common that they use OMyMakefile. Could it be that the bug is there? OMyMakefile replaces many of the build macros coming with omake.

A workaround could be to run omake with `-j 1` to prevent any parallel execution.

**Comment by rleonid** [comment]
I'm not certain if it is related (assumptions over interface vs implementations) but, I [cfstream](https://github.com/biocaml/cfstream) reliably fails to install on some travis tests that I run: https://travis-ci.org/hammerlab/prohlatype/jobs/195985901. Only OSX.

**Comment by rgrinberg** [comment]
@rleonid would it be possible for you to experiment with the -j1 flag and see if that helps anything?

**Comment by gerdstolpmann** [comment]
So, I could reproduce the problem on OSX - but not on Linux. One of the few differences between the two OSes is that `Unix.stat` returns high-resolution mtimes on Linux but only second-precision mtimes on OSX. That could have an effect on the internal cache of omake. I'm now checking the code.

**Comment by gerdstolpmann** [comment]
Note that the problem is still rare on OSX. A repeated build took 29-75 rounds of "omake clean; omake" until it ran into the problem.

**Comment by gerdstolpmann** [comment]
It turns out this has to do with the case-insensitive file system on Macs. E.g. for cfstream I could see that both `CFStream.mli` and `cFStream.mli` were compiled. That interfered with a parallel running build of a descendent of that module, as the cmi file was overwritten while the descendent was being compiled.

This fixed it: https://github.com/ocaml-omake/omake/pull/79

**Comment by gerdstolpmann** [comment]
As far as the badly implemented case sensitivity check on Macs is concerned, this problem is fixed. Closing because of this. If any more instability is observed please create new ticket.
